### DISCUSSION TITLE
Gists webhooks

### DISCUSSION THREAD

**Discussion by eerison** [community/community#50313] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Hello is there anyway or plan to create events/hooks for gists?

I would like to trigger pipelines when create or update a gist.

**Comment by marnovandermaas** [comment]
There is a REST API for managing gists: https://docs.github.com/en/rest/gists/gists?apiVersion=2022-11-28
And also for managing gist comments: https://docs.github.com/en/rest/gists/comments?apiVersion=2022-11-28
So you could detect creation and updates of gists and trigger a pipeline based on that.

**Reply by eerison** [reply]
But in this case I need to request the gist repo time to time to see the new updates or creations.

and it doesn't help me, because I need to save the gists url or users url to request, and if it grows it will be a lot of requests and most of them are unnecessary.

**Comment by Mohamed-SayedAlAhl** [comment, marked as answer]
you can use the API endpoints for listing, creating, updating, and deleting Gists and their comments. You can then use a webhook to receive notifications from GitHub when a Gist is created or updated, and trigger your pipeline accordingly.

Here is an example workflow:

Set up a webhook to receive notifications from GitHub when a Gist is created or updated.
When a webhook is triggered, use the GitHub API to retrieve information about the Gist.
Determine whether the Gist is a new Gist or an updated version of an existing Gist.
If the Gist is new, trigger your pipeline to process the Gist.
If the Gist is an update, trigger your pipeline to update the previously processed Gist.
Remember to follow best practices when using webhooks, such as verifying the signature of the webhook payload to ensure its authenticity.

Alternatively, you can use a third-party service such as Zapier or IFTTT to monitor your Gists for changes and trigger a pipeline when a new Gist is created or updated.

Keep in mind that there may be rate limits and other limitations on how frequently you can poll the GitHub API, so it's important to design your implementation accordingly.

**Reply by eerison** [reply]
> You can then use a webhook to receive notifications from GitHub when a Gist is created or updated, and trigger your pipeline accordingly.

do you have any example how to create this webhook.

**Reply by Mohamed-SayedAlAhl** [reply]
Yes, here's an example of how you can create a webhook for Gist events in GitHub using the command line:

1. Generate a personal access token with the `gist` scope enabled in your GitHub account settings.

2. Use the following command to create a webhook:

```
curl -i -H "Authorization: token YOUR_PERSONAL_ACCESS_TOKEN" -H "Content-Type: application/json" -X POST -d '{"name": "web", "active": true, "events": ["gist"], "config": {"url": "YOUR_WEBHOOK_URL", "content_type": "json", "secret": "YOUR_WEBHOOK_SECRET"}}' https://api.github.com/repos/YOUR_USERNAME/YOUR_REPOSITORY/hooks
```

Replace `YOUR_PERSONAL_ACCESS_TOKEN` with the personal access token you generated in step 1, `YOUR_WEBHOOK_URL` with the URL of the endpoint where you want to receive webhook events, `YOUR_WEBHOOK_SECRET` with a secret string that you will use to verify the authenticity of the webhook payload, `YOUR_USERNAME` with your GitHub username, and `YOUR_REPOSITORY` with the name of the repository where you want to create the webhook.

This command creates a webhook that will be triggered for Gist events and sends the payload to the specified endpoint.

3. Once you have created the webhook, you can use it to receive notifications for Gist events and trigger your pipeline accordingly. You can use the GitHub API to retrieve information about the Gist and determine whether it's a new Gist or an updated version of an existing Gist, as described in the previous answer.

**Reply by eerison** [reply]
Thank you,  I'll try that  :)

**Reply by cupnoodle** [reply]
@eerison did the example work for you? What value do you use for the repo name?

**Reply by eerison** [reply]
Hey @cupnoodle , to be honest,  I haven't tried

**Reply by 0xdevalias** [reply]
I don't believe this works.. webhooks for an individual repo don't seem to have scopes for seeing gists (which are account level), and there doesn't seem to be a way to subscribe to account level webhooks; seemingly only for organizations/etc; but even that doesn't seem to have an option for gists:

- https://docs.github.com/en/webhooks/using-webhooks/creating-webhooks

**Reply by vimagick** [reply]
```json
{
  "message": "Validation Failed",
  "errors": [
    {
      "resource": "Hook",
      "code": "custom",
      "message": "gist is not a valid event name"
    }
  ],
  "documentation_url": "https://docs.github.com/rest/repos/webhooks#create-a-repository-webhook",
  "status": "422"
}
```

**Comment by Pasan-Pahasara** [comment]
Begin by configuring a webhook that listens for events from GitHub whenever a Gist is created or modified.

Upon receiving a webhook notification, utilize the GitHub API to gather details about the Gist that triggered the event.

Analyze whether the Gist is a brand-new one or an update to an existing Gist.

If the Gist is new, initiate your pipeline to handle and process the newly created Gist.

If the Gist is an update to an existing Gist, trigger your pipeline to make necessary updates to the previously processed Gist.

It's important to adhere to best practices when working with webhooks, such as verifying the payload's signature to ensure its legitimacy.

As an alternative approach, you can employ third-party services like Zapier or IFTTT to actively monitor your Gists for changes and then automatically trigger your pipeline when a new Gist is created or an existing one is updated.

Keep in mind that the GitHub API may have rate limits and other restrictions, so it's crucial to plan your implementation carefully to avoid hitting these limits and ensure smooth automation.

**Reply by 0xdevalias** [reply]
> I don't believe this works.. webhooks for an individual repo don't seem to have scopes for seeing gists (which are account level), and there doesn't seem to be a way to subscribe to account level webhooks; seemingly only for organizations/etc; but even that doesn't seem to have an option for gists:
> 
> - https://docs.github.com/en/webhooks/using-webhooks/creating-webhooks
> 
> _Originally posted by @0xdevalias in https://github.com/orgs/community/discussions/50313#discussioncomment-7570477_

**Comment by cupnoodle** [comment]
Hi Pasan,

Please show code example and the exact webbook topic used, I searched
Github documentations for REST API for OAuth / Github Apps , theres no
webhook topic that allow us to listen to Gist create / update.
